### ISSUE TITLE
A proposal for "image sets" - how to deal with differing aspect ratios between sets and their automated creation

### ISSUE THREAD

**Issue by rpkoller** [ResponsiveImagesCG/wp-tevko-responsive-images#45]
Hi,
I was forwarded over here from the Slack #feature-respimg channel as the recommended place to suggest feature proposals regarding the responsive image plugin.
So far i was using responsive images with the Advanced Custom Fields plugins image object functionality and the Picturefill plugin (version 1.x). The designs I use contain `x1.5` and `x2` retina images as well as different sets of images with varying aspect ratios on differing break points. Which means i have normally e.g. about three to four sets with a number of widths for each at certain breakpoints. One set alone would hold e.g. about eight versions per image in sum.  A few sizes are available in every set. But even without those I normally still end up with 15 to 30 images in my projects in the end. Instead of being able to adjust the content image creation to the exact amount of sizes needed for a particular images you end up with too many unnecessary image versions as well as tenth of unneeded megabytes in sum.
In the following I split the text into two parts. At first the present state is described, followed by the suggested solution afterwards. I have to add that I kept the underlying constraint in consideration that things should be kept simple for the end user while it should be still possible for developers to configure and tweak on a per theme basis via the functions.php file. I hope I’ve headed towards the happy medium. 

## Present State
At the moment I’ve normally used the following steps. First I’ve unsetted the default wp sizes 

```php
function df_remove_default_image_sizes( $sizes ) {
	unset( $sizes['thumbnail']);
	unset( $sizes['medium']);
	unset( $sizes['large']);
	return $sizes;
}
add_filter( 'intermediate_image_sizes_advanced', 'df_remove_default_image_sizes' );
```

In the next step all necessary image sizes are added (I’ve listed only one example per resolution). 

```php
add_image_size('is268', 268, 9999);
add_image_size('is268_x15', 402, 9999);
add_image_size('is268_x2', 536, 9999);
```
Depending on the layout you might end up with 15 to 30 versions per image over all, sometimes even more. But for example a content image in a narrow sidebar never reaches broader widths. Therefor only a fraction of the created images versions would be necessary. 

### The Constraints
As I said before the constraints are the following: 
* Don't swamp the enduser with too many options
* Leave the developers enough hooks and configuration options to adjust things to their needs
* The configuration should be on a per theme basis in the functions.php and not via the global user interface. 

### The Problem
* There are different content images with differing aspect ratios in differing breakpoint scenarios ( feature images, narrow images in sidebars etc.) 
* Possible that you end up with three to four or even more different types/sets of images. 
* As described in the „present state“ section you end up with tenth of different images per image to cover all sets and scenarios. 
* But basically the sidebar example would only need e.g. six.

### The Proposed Solution
* In the WP Admin interface the enduser is able to adjust the extents of the three default media sizes. (Probably enough in a lot of use cases in combination with the img src method.)
* I would entitle it the default „image set“
* In the functions.php it should be possible to modify that default image set too: 
   * Remove or modify the existing sizes from the admin interface
    * Add more sizes
    * Add actions that retina images should be generated for the particular image set. Which means if e.g. x2 is set, WordPress would automatically generate twice the width and height of the set sizes. No extra configuration and calculation needed from the developer side (x1.5 and x3 would be analogous if set) 
    * Note: I am unsure if it would make sense to reflect the functions.php settings of the default image set directly into the media preferences inside the admin interface or leave it there as is at the moment. 
* Aside the default image set you would be able to create and name more image sets
    * Inside each set you would be able to set widths and heights like in the default set
    * The retina settings get adopted for each image set from the default parent image set. 
* As soon as there are more image sets aside the default one, the image edit inside the media center should show up a select field showing the list of available image sets. So you are able to reassign an image to a certain image set. 
    * Per default new uploaded images are assigned to the default image set and the sizes of the default image set are created
    * If you change the image set from default to e.g. the one named sidebar, when hitting ok, the old sizes are deleted and the sizes of the selected image set are created. 

That is just the rough outline. I haven't gone into detail. Just wanted to outline the current flaw of image creation in WordPress in regard of responsive images. The present state might be ok for a site with 5 or 10 images. But as soon as you have a more complex site with many images over the time the amount of unnecessary images never used would rise exponentially as well as the need for disk space. You might prevent that situation with an approach like the proposed image sets. Any thoughts? ;) Cheers Ralf

**Comment by joemcgill** [comment]
Thanks for the detailed explanation @rpkoller. If I'm understanding correctly, it sounds like you're proposing an API that could be used by theme authors to expose a UI in the admin for managing sets of images, which could then be used for building responsive image sets.

1. Do I have the basic idea right?
2. Could a theme author use current WordPress functions to provide this functionality or would this require modifications to core?

**Comment by rpkoller** [comment]
1. Roughly yes. Showing off and make ALL sets accessible inside the admin interface as soon they are defined inside the functions.php i've ruled out until now since i've thought that things should kept simple in the front end. In my initial proposal only the default set ( the initial sizes or the changed ones should have been shown). But for the further tweaking it would be handy and convenient if also the other defined sets would show up. Might be confusing if only the default set would show up in the backend and the rest does not even if you it set up inside the functions.php. But on the other hand i have to admit, for the regular enduser, it might be confusing too if a theme author has set several sets and the user now sees those sets instead of the regular three sizes he or she is used to. But i think it would be useful nevertheless as soon as any set is registered that it is displayed inside the media preferences anyway - would be a more consequent move. 
And it wouldn't be just managing and defining sizes alongside the display in the media preferences. You would also need a select box for the image set in the media edit screen that you are able to assign a set of sizes to a single image. And in the end WordPress has to replace the default size for an image with the appropriate sizes necessary reflecting the set image set (or update). 

2. With the current WP functions you are only able to create the described "default set" half ways,  which means unset  the default three sizes and replace them with a set of image sizes you need. Which means you need to merge all sizes necessary for all sets into one. Which leads to the problem of too many unnecessary images i initially descibed. And also the size changes aren't reflected back to the media preferences. And anyway WordPress doesn't KNOW different image sets at all so far.

So to sum things up,a bit of underlying functionality is already there but i suppose there would be the need of a few modifications and extensions to the core. But in my humble opionen it would be more than benificial to have flexible responsive images managing functionality aside the orchestration with the present state of the plugin.

**Comment by tevko** [comment]
What we cant do is add an image interface to the wordpress admin. I think that will only make the plugin more complicated while leading us further away from our original goal: default responsive images that work invisibly. 

Whatever solution we come up with, it will have to be admin-less, while allowing theme developers to customize and improve as needed. We won't be able to cover every possible use case and that's okay

**Comment by rpkoller** [comment]
@tevko yep would be fine with me too. The only necessary ui element in my suggested scenario would be the select element inside the media edit screen ( as soon as more than one image set would be defined inside the functions.php) so that the user can apply an image to another image set. That would be the most unobtrusive way i could imagine.

**Comment by tevko** [comment]
Hey @rpkoller , while you've very interesting issue, we've determined it to be just a bit outside of the scope what we're trying to achieve with this plugin. Please continue to contribute, as we value your ideas and insight. For the time being however, we're going to close this and move on to more pressing issues
